In [1]:
import random

# 生成50000个32位随机整数
# Генерируем 50 000 32-битных случайных целых чисел
numbers = [random.randint(0, 2**32 - 1) for _ in range(50000)]

# 将这些数字写入文件，每个数字占一行
# Запишите эти числа в файл, по одной строке на каждое число
with open('numbers.txt', 'w') as f:
    for number in numbers:
        f.write(str(number) + '\n')

In [2]:
import time

def prime_factors(n):
    i = 2
    factors = []
    while i * i <= n:
        if n % i:
            i += 1
        else:
            n //= i
            factors.append(i)
    if n > 1:
        factors.append(n)
    return factors

start_time = time.time()

# 读取文件中的数字
# Чтение чисел из файла
with open('numbers.txt', 'r') as f:
    numbers = [int(line.strip()) for line in f]

# 计算所有数字的素数因子总数
# Подсчитайте общее количество простых факторов для всех чисел
total_factors = sum(len(prime_factors(n)) for n in numbers)

end_time = time.time()
running_time = end_time - start_time

print(f"Total factors: {total_factors}")
print(f"Running time: {running_time} seconds")

Total factors: 206138
Running time: 18.693769931793213 seconds


In [4]:
import ray
import time

ray.init()

@ray.remote
def prime_factors(n):
    i = 2
    factors = []
    while i * i <= n:
        if n % i:
            i += 1
        else:
            n //= i
            factors.append(i)
    if n > 1:
        factors.append(n)
    return factors

start_time = time.time()

# 读取文件中的数字
# Чтение чисел из файла
with open('numbers.txt', 'r') as f:
    numbers = [int(line.strip()) for line in f]

# 使用Ray并行计算所有数字的素数因子总数
# Вычислите общее количество простых факторов для всех чисел параллельно, используя Ray
results = ray.get([prime_factors.remote(n) for n in numbers])
total_factors = sum(len(result) for result in results)

end_time = time.time()
running_time = end_time - start_time

print(f"Total factors: {total_factors}")
print(f"Running time: {running_time} seconds")

ray.shutdown()

2024-04-30 19:45:30,993	INFO worker.py:1752 -- Started a local Ray instance.


Total factors: 206138
Running time: 4.85775899887085 seconds


- многопоточно (на CPython - multiprocessing), с использованием примитивов синхронизации   
Поскольку у меня возникли проблемы с модулем мультипроцессинга в Jupyter notebook, я создал отдельный python-файл    
исходный код : task2_multiprocessing.py    
Total prime factors: 206138    
Processing time: 1.55 seconds